flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/DenseVector.scala

```diff
@@ -102,6 +102,38 @@ case class DenseVector(
     }
   }
 
+  /** Returns the outer product (a.k.a. Kronecker product) of `this`
+    * with `other`. The result will given in [[org.apache.flink.ml.math.SparseMatrix]]
+    * representation if `other` is sparse and as [[org.apache.flink.ml.math.DenseMatrix]] otherwise.
+    *
+    * @param other a Vector
+    * @return the [[org.apache.flink.ml.math.Matrix]] which equals the outer product of `this`
+    *         with `other.`
+    */
+  override def outer(other: Vector): Matrix = {
+    val numRows = size
+    val numCols = other.size
+
+    other match {
+      case sv @ SparseVector(_, _, _) =>
+        val entries = for {
+          i <- 0 until numRows
+          j <- sv.indices
+          value = this(i) * sv.data(sv.indices.indexOf(j))
```

Comment:
`sv.indices.indexOf(j)` will in the best case always trigger a binary search. Why not doing `(j, idxJ) <- sv.indices.zipWithIndex` and then using `idxJ` here?